web/README.md

```diff
@@ -0,0 +1,42 @@
+# Web Stuff
+
+This is a collection of API's and build tools for packaging our various
+Widgetry apps as web applications.
+
+## Goals
+
+A web developer, who might not know anything about rust or wasm, should be able
+to use our packaged javascript libraries on their website with minimal
+customization or arcanery.
+
+Users of their website should be able to interact with the widgetry app without
+it feeling weird or having to jump through hoops.
+
+## Limitations
+
+### JS feature: `import.meta`
+
+To allow the application to live at any URL (rather than presupposing it lives
+at root, or whatever), we rely on `import.meta` which isn't supported on some
+browsers before 2018. See: https://caniuse.com/?search=import.meta
+
+An alternative would be to require configuration, so the loader knows where to
+download it's "*_wasm.bg file".
+
+### Browser Feature: WebGL
+
+We prefer WebGL2, but now gracefully fall back to WebGL1. This should cover
+all common browsers since late 2014. https://caniuse.com/?search=webgl
+
+## Examples
+
+See [`examples/*.js`](examples/) for code examples.
+
+You can build and see the examples in your webbrowser with:
+
+```
+// install typescript build dependency
```

Comment:
I'm not familiar with npm usage, but I have it installed and ran `npm install` successfully. `make build` fails because there's no `tsc` command in my `$PATH`. I see `package.json` depends on typescript, so do I need to separately install it?

Reply:
`apt-get install node-typescript` got me further, but now `make build` says:

```
tsc --lib dom,es2020 --module es2020 --target es6 --strict --noEmitOnError --outDir build src/widgetry.ts
error TS6046: Argument for '--lib' option must be: 'es5', 'es6', 'es2015', 'es7', 'es2016', 'es2017', 'es2018', 'esnext', 'dom', 'dom.iterable', 'webworker', 'scripthost', 'es2015.core', 'es2015.collection', 'es2015.generator', 'es2015.iterable', 'es2015.promise', 'es2015.proxy', 'es2015.reflect', 'es2015.symbol', 'es2015.symbol.wellknown', 'es2016.array.include', 'es2017.object', 'es2017.sharedmemory', 'es2017.string', 'es2017.intl', 'es2017.typedarrays', 'esnext.array', 'esnext.asynciterable', 'esnext.promise'.
error TS6046: Argument for '--module' option must be: 'none', 'commonjs', 'amd', 'system', 'umd', 'es6', 'es2015', 'esnext'.
Makefile:79: recipe for target 'build/widgetry.js' failed
make: *** [build/widgetry.js] Error 1
```

Reply:
whoops - I've pushed up a commit that should fix this.

You can uninstall that system tsc - it should use the one in node_modules now, which should have been installed when running npm install.

Reply:
Thanks! Some more questions as I'm running this:

1) There are some warnings from `make build`:

```
rm src/fifteen_min/wasm_pkgmake: unlink: src/fifteen_min/wasm_pkg: Is a directory
 src/santa/wasm_pkgmake: unlink: src/santa/wasm_pkg: Is a directory
 src/widgetry_demo/wasm_pkgmake: unlink: src/widgetry_demo/wasm_pkg: Is a directory
 src/osm_viewer/wasm_pkgmake: unlink: src/osm_viewer/wasm_pkg: Is a directory
```

I see the `clean` rule does `rm -r`, but I'm not sure where the rest of this unlinking is happening.

2) After running, `package-lock.json` has been modified to remove the `packages` dict entirely, and set `lockfileVersion` to 1. There was a warning about an old npm, so maybe I need to first upgrade.

3) I'm not figuring out how to get the server started after building:

```
[5][~/abstreet/web]$ make serve
make: *** No rule to make target 'serve'.  Stop.
[5][~/abstreet/web]$ make server
python -m http.server 8000 --directory build/dist
[5][~/abstreet/web]$ make run
echo "Make sure server is up and running first!"
Make sure server is up and running first!
open http://localhost:8000
Couldn't get a file descriptor referring to the console
Makefile:24: recipe for target 'run' failed
make: *** [run] Error 
```

Reply:
> open http://localhost:8000

Hmm... that might be a macos-ism... 

Is there a command line to "open a browser to the target url"?

IIRC maybe it's called "xdg-open"

Reply:
> After running, package-lock.json has been modified to remove the packages dict entirely, and set lockfileVersion to 1. There was a warning about an old npm, so maybe I need to first upgrade.

I'm guessing you have a different version of npm installed.

I'm not a javascript pro, but my understanding is that the lock shouldn't be modified unless you're running something like `npm install` or `npm update`. 

Locally I'm at:
```
$ npm --version
7.6.3
```

We could introduce a .node-version and require nodeenv as prerequisite, but I'd just as soon avoid that developer hurdle if the version mismatch is not actually causing problems. Can you revert the changes and verify that they were a one-time thing related to `npm install`, or do they happen every time you build?

Reply:
> There are some warnings from make build:

Fixed!

Reply:
> Can you revert the changes and verify that they were a one-time thing related to npm install, or do they happen every time you build?

Ah, it's just `npm install`. Things're fine as they are. I have npm 6.4.16 somehow, and the warning is `This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!`. No worries.

> IIRC maybe it's called "xdg-open"

Indeed! I think we can do something like this:

```
if [ "$(uname)" == "Linux" ]; then
  BROWSER=xdg-open
else
  BROWSER=open
fi
```

I've almost got things going, but
1) The README is a bit wrong -- there is no `make serve`. Is it supposed to be `make server` in one shell, and `make run` in another?
2) The `python -m http.server` command instantly exits on my machine with no logs or status code. `python3 -m http.server 8000` works, but it doesn't seem to have a `--directory` option. I'm a bit concerned the dependencies and process for running wasm locally are getting more complicated / system-dependent, although not many people will be doing this.

Reply:
> I'm a bit concerned the dependencies and process for running wasm locally are getting more complicated / system-dependent

We've added a prerequisite of npm and gnu make, which seem pretty reasonable to me.

I don't know that the "more system dependent" statement is totally fair. I suspect it's at least partially because historically you've written most of the code, so "system independent" vs "works on my machine" is less  distinguishable. Believe me, I've spent plenty of time trying to make abstreet and various rust libraries in general work on my system.

I think you're mostly hitting rough edges that we can trivially smooth out.

There is some amount of complexity added with the makefile. I could remove it all and have a oneshot script that does everything, but I prefer the ergonomics of being able to do `make widgetry_demo` without having to edit that script.

> The README is a bit wrong -- there is no make serve. Is it supposed to be make server in one shell, and make run in another?

Whoops - typo. Yes, `make server` and `make run` in another. I'm just going to delete the `make run` stuff since it's really not doing that much and it's become a source of friction.

> The python -m http.server command instantly exits on my machine with no logs or status code. python3 -m http.server 8000 works, but it doesn't seem to have a --directory option.

Seems like you're using an old version of http.server and apparently you're on a system which is still clinging to maintain a python2 installation.

I think we can make this work on machines with older python versions with:
```diff
- python -m http.server --directry $foo
+ cd $foo && python3 -m http.server
```

Reply:
> Believe me, I've spent plenty of time trying to make abstreet and various rust libraries in general work on my system.

True -- I forgot about the gdal version mess, and some of the early Mac issues with number of OpenGL textures. I feel s
poiled by how self-contained Cargo is, and I want to keep up the easy-to-install/use spirit of abst generally.

> on a system which is still clinging to maintain a python2 installation.

Thanks for the patience with my weird Ubuntu 18.04 versions -- I'm happy with the flexibility of the makefile, and no longer skeptical about the new dependencies.

Reply:
I definitely favor keeping things easy to install and use.

Thanks for the feedback and helping shave off some of the rough edges.